### PR TITLE
Distributed alter query(delete/update/drop partition) on cross replication clusters

### PR DESCRIPTION
**Pull request by etah000**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Short description
Add support for distributed ddl(update/delete/drop partition) on crosss replication clusters.

Detailed description (optional):
For distributed ddl on crosss replication clusters, the syntax should be similar to the following "drop" statement:
`alter table db1.tab1_local on cluster cluster1 drop partition '2018-01-01';`
On the other types of clusters, the DDLs remain the same as before.

### CODE REVIEW

File: src/Interpreters/DDLWorker.cpp [hunk 1]
```diff
@@ -494,14 +495,30 @@ void DDLWorker::parseQueryAndResolveHost(DDLTask & task)
...
                 {
-                    throw Exception("There are two exactly the same ClickHouse instances " + address.readableString()
-                        + " in cluster " + task.cluster_name, ErrorCodes::INCONSISTENT_CLUSTER_DEFINITION);
+                    if(default_database == address.default_database)
+                    {
+                        throw Exception(
+                            "There are two exactly the same Snowball instances " + address.readableString() + " in cluster "
+                                + task.cluster_name, ErrorCodes::INCONSISTENT_CLUSTER_DEFINITION);
+                }
+                    else  ///circular replication is used.
```
Comment (nikitamikhaylov) [inline src/Interpreters/DDLWorker.cpp:504]: What does mean `circular replication`?

Comment (nvartolomei) [reply src/Interpreters/DDLWorker.cpp:504]: https://www.altinity.com/blog/2018/5/10/circular-replication-cluster-topology-in-clickhouse

**Comment by nvartolomei** [comment]
Related #1287?

**Comment by etah000** [comment]
For my test case in integration test, it reported "RequestsDependencyWarning: urllib3 (1.23) or chardet (3.0.4) doesn't match a supported version!", this is related with "requests" package in python. 

For other failed tests, I can see no relation with the changes I made, so what shall I do next?

**Comment by nikitamikhaylov** [comment]
Hmm, you checkout the branch from 20.4 and try to merge into 20.4. I will fix.

**Comment by nikitamikhaylov** [comment]
Your commits now live here #11703. I will probably close this PR.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -4072,10 +4072,10 @@ void StorageReplicatedMergeTree::sendRequestToLeaderReplica(const ASTPtr & query
 
     /// SECONDARY_QUERY here means, that we received query from DDLWorker
     /// there is no sense to send query to leader, because he will receive it from own DDLWorker
-    if (query_context.getClientInfo().query_kind == ClientInfo::QueryKind::SECONDARY_QUERY)
-    {
-        throw Exception("Cannot execute DDL query, because leader was suddenly changed or logical error.", ErrorCodes::LEADERSHIP_CHANGED);
-    }
+//    if (query_context.getClientInfo().query_kind == ClientInfo::QueryKind::SECONDARY_QUERY)
```
Comment (nikitamikhaylov) [inline src/Storages/StorageReplicatedMergeTree.cpp:4075]: Is this necessary?

**Comment by azat** [comment]
@nikitamikhaylov looks like you pushed to the `refs/heads/pr/11508` remote branch in the upstream repository, and it conflicts with other PR's refs, will it be rename/delete eventually? Or should I change the local ref folder?

```
$ git fetch upstream 
fatal: Cannot fetch both refs/heads/pr/11508 and refs/pull/11508/head to refs/remotes/upstream/pr/11508
```

**Comment by nikitamikhaylov** [comment]
@azat Yes, I pushed empty commit here and I don’t know how to return it back :) Maybe we can simply close this PR and corresponding branch will be deleted?

**Comment by azat** [comment]
>Maybe we can simply close this PR and corresponding branch will be deleted?

This will not helps, since you did not push the ref of PR but create new one:
- PR ref - `refs/pull/11508/head`
- You pushed to ref - `refs/heads/pr/11508`

To remove create ref by your you can run the following:

```
git push upstream :pr/11508
```

**Comment by azat** [comment]
Or better using qualified path:

```
git push upstream :refs/heads/pr/11508
```

**Comment by nikitamikhaylov** [comment]
> To remove create ref by your you can run the following:
> 
> ```
> git push upstream :pr/11508
> ```
Done

**Pull request closed without merging**